sn_client/src/audit/spend_dag_building.rs

```diff
@@ -213,39 +217,58 @@ impl Client {
             // only verify parents we haven't already verified
             txs_to_verify = next_gen_tx
                 .into_iter()
-                .filter(|tx| !verified_tx.contains(&tx.hash()))
+                .filter(|tx| !known_txs.contains(&tx.hash()))
                 .collect();
 
             depth += 1;
             let elapsed = start.elapsed();
-            let n = verified_tx.len();
-            info!("Now at depth {depth} - Verified {n} transactions in {elapsed:?}");
+            let n = known_txs.len();
+            info!("Now at depth {depth} - Collected spends from {n} transactions in {elapsed:?}");
         }
 
         let elapsed = start.elapsed();
-        let n = verified_tx.len();
-        info!("Verified all the way to known spends or genesis! Through {depth} generations, verifying {n} transactions in {elapsed:?}");
+        let n = known_txs.len();
+        info!("Collected the DAG branch all the way to known spends or genesis! Through {depth} generations, collecting spends from {n} transactions in {elapsed:?}");
+
+        // verify the DAG
+        info!("Now verifying SpendDAG extended at {spend_addr:?} and recording errors...");
+        let start = std::time::Instant::now();
+        if let Err(e) = dag.record_faults(&dag.source()) {
+            let s = format!(
+                "Collected DAG starting at {spend_addr:?} is invalid, this is probably a bug: {e}"
+            );
+            error!("{s}");
+            return Err(WalletError::Dag(s));
+        }
+        let elapsed = start.elapsed();
+        info!("Finished verifying SpendDAG extended from {spend_addr:?} in {elapsed:?}");
         Ok(())
     }
 
     /// Extends an existing SpendDag starting from the utxos in this DAG
     /// Covers the entirety of currently existing Spends if the DAG was built from Genesis
+    /// Records errors in the new DAG branches if any
     pub async fn spend_dag_continue_from_utxos(&self, dag: &mut SpendDag) -> WalletResult<()> {
-        info!("Gathering spend DAG from utxos...");
+        trace!("Gathering spend DAG from utxos...");
         let utxos = dag.get_utxos();
-        let mut tasks = JoinSet::new();
-        for utxo in utxos {
-            info!("Launching task to gather utxo: {:?}", utxo);
-            let self_clone = self.clone();
-            tasks.spawn(async move { self_clone.spend_dag_build_from(utxo).await });
-        }
-        while let Some(res) = tasks.join_next().await {
-            let sub_dag = res.map_err(|e| {
-                WalletError::FailedToGetSpend(format!("DAG gathering task failed: {e}"))
-            })??;
-            dag.merge(sub_dag);
+
+        let tasks: Vec<_> = utxos
+            .iter()
+            .map(|utxo| {
+                debug!("Launching task to gather DAG from utxo: {:?}", utxo);
+                self.spend_dag_build_from(*utxo)
+            })
+            .collect();
```

Comment:
Just a tiny suggestion, but we can probably use `futures stream` here to run atmost `n` tasks concurrently. I feel the utxos here can get very big and we might not want to spawn tasks for all of them.


Here is an example snippet
```rust
        let mut stream = futures::stream::iter(utxos.into_iter())
            .map(|utxo| self.spend_dag_build_from(utxo))
            .buffer_unordered(MAX_CONCURRENT_TASKS);


        while let Some(result) = stream.next().await {
			}
```